File: sentinel_media_publisher/publisher.js
const fs = require('fs');
const { google } = require('googleapis');
const path = require('path');
const readline = require('readline');

// Configuración
const SCOPES = ['https://www.googleapis.com/auth/youtube.upload'];
const TOKEN_PATH = path.join(__dirname, 'token.json');
const CREDENTIALS_PATH = path.join(__dirname, '../client_secrets.json');

async function authorize(credentials) {
  const { client_secret, client_id, redirect_uris } = credentials.installed || credentials.web;
  const oAuth2Client = new google.auth.OAuth2(client_id, client_secret, redirect_uris[0]);

  try {
    const token = fs.readFileSync(TOKEN_PATH);
    oAuth2Client.setCredentials(JSON.parse(token));
    return oAuth2Client;
  } catch (err) {
    return getNewToken(oAuth2Client);
  }
}

function getNewToken(oAuth2Client) {
  return new Promise((resolve, reject) => {
    const authUrl = oAuth2Client.generateAuthUrl({ access_type: 'offline', scope: SCOPES });
    console.log('Authorize this app by visiting this url:', authUrl);
    const rl = readline.createInterface({ input: process.stdin, output: process.stdout });
    rl.question('Enter the code from that page here: ', (code) => {
      rl.close();
      oAuth2Client.getToken(code, (err, token) => {
        if (err) return reject('Error retrieving access token', err);
        oAuth2Client.setCredentials(token);
        fs.writeFileSync(TOKEN_PATH, JSON.stringify(token));
        resolve(oAuth2Client);
      });
    });
  });
}

async function uploadVideo(auth, filePath, title, description, privacy) {
  const service = google.youtube('v3');
  const fileSize = fs.statSync(filePath).size;
  
  console.log(`🚀 Uploading: ${title} (${privacy})`);
  
  const res = await service.videos.insert({
    auth: auth,
    part: 'snippet,status',
    requestBody: {
      snippet: {
        title: title,
        description: description,
        tags: ['Sentinel', 'AI', 'NodeJS'],
        categoryId: '28',
      },
      status: {
        privacyStatus: privacy,
      },
    },
    media: {
      body: fs.createReadStream(filePath),
    },
  }, {
    onUploadProgress: evt => {
      const progress = (evt.bytesRead / fileSize) * 100;
      process.stdout.write(`\rProgress: ${Math.round(progress)}%`);
    },
  });
  
  console.log('\n✅ Upload complete!');
  console.log(`https://youtu.be/${res.data.id}`);
}

const args = process.argv.slice(2);
const fileArg = args.find(a => a.startsWith('--file='));
const titleArg = args.find(a => a.startsWith('--title='));
const descArg = args.find(a => a.startsWith('--desc='));
const privacyArg = args.find(a => a.startsWith('--privacy=')) || '--privacy=private';

if (!fileArg) {
  console.error("Usage: node publisher.js --file=VIDEO.mp4 [--title=TITLE] [--desc=DESC] [--privacy=private]");
  process.exit(1);
}

const filePath = fileArg.split('=')[1];
const title = titleArg ? titleArg.split('=')[1] : path.basename(filePath, path.extname(filePath));
const desc = descArg ? descArg.split('=')[1] : "Uploaded by Sentinel Node Publisher";
const privacy = privacyArg.split('=')[1];

fs.readFile(CREDENTIALS_PATH, (err, content) => {
  if (err) return console.log('Error loading client secret file:', err);
  authorize(JSON.parse(content))
    .then(auth => uploadVideo(auth, filePath, title, desc, privacy))
    .catch(console.error);
});

File: sentinel_media_publisher/publisher.py
#!/usr/bin/env python3
import os
import sys
import argparse
import pickle
import json
from google_auth_oauthlib.flow import InstalledAppFlow
from google.auth.transport.requests import Request
from googleapiclient.discovery import build
from googleapiclient.http import MediaFileUpload

# SCOPES para YouTube Data API v3
SCOPES = ['https://www.googleapis.com/auth/youtube.upload']
CLIENT_SECRETS_FILE = os.path.join(os.path.dirname(os.path.dirname(os.path.abspath(__file__))), 'client_secrets.json')
TOKEN_PICKLE_FILE = 'token.pickle'

def get_authenticated_service():
    creds = None
    if os.path.exists(TOKEN_PICKLE_FILE):
        with open(TOKEN_PICKLE_FILE, 'rb') as token:
            creds = pickle.load(token)
    
    if not creds or not creds.valid:
        if creds and creds.expired and creds.refresh_token:
            creds.refresh(Request())
        else:
            if not os.path.exists(CLIENT_SECRETS_FILE):
                print(f"❌ Error: {CLIENT_SECRETS_FILE} not found.")
                sys.exit(1)
            
            flow = InstalledAppFlow.from_client_secrets_file(CLIENT_SECRETS_FILE, SCOPES)
            creds = flow.run_local_server(port=0)
        
        with open(TOKEN_PICKLE_FILE, 'wb') as token:
            pickle.dump(creds, token)

    return build('youtube', 'v3', credentials=creds)

def upload_video(file_path, title, description, privacy, category_id="28"):
    youtube = get_authenticated_service()

    body = {
        'snippet': {
            'title': title,
            'description': description,
            'tags': ['Sentinel', 'AI', 'Rust', 'Research'],
            'categoryId': category_id
        },
        'status': {
            'privacyStatus': privacy,
            'selfDeclaredMadeForKids': False,
        }
    }

    print(f"🚀 Uploading {file_path}...")
    media = MediaFileUpload(file_path, chunksize=-1, resumable=True)
    request = youtube.videos().insert(part=','.join(body.keys()), body=body, media_body=media)

    response = None
    while response is None:
        status, response = request.next_chunk()
        if status:
            print(f"   Upload progress: {int(status.progress() * 100)}%")

    print(f"✅ Upload Complete! Video ID: {response.get('id')}")
    print(f"   URL: https://youtu.be/{response.get('id')}")

def main():
    parser = argparse.ArgumentParser(description='Sentinel Publisher (Python Engine)')
    parser.add_argument('--file', required=True, help='Path to video file')
    parser.add_argument('--title', help='Video Title')
    parser.add_argument('--description', help='Video Description')
    parser.add_argument('--privacy', default='private', choices=['public', 'private', 'unlisted'], help='Privacy Status')
    
    args = parser.parse_args()

    if not os.path.exists(args.file):
        print(f"❌ Error: File {args.file} not found.")
        sys.exit(1)

    # Auto-detect title from filename if missing
    title = args.title if args.title else os.path.splitext(os.path.basename(args.file))[0]
    description = args.description if args.description else "Uploaded via Sentinel Publisher"

    try:
        upload_video(args.file, title, description, args.privacy)
    except Exception as e:
        print(f"❌ An error occurred: {e}")
        sys.exit(1)

if __name__ == '__main__':
    main()
